### PR TITLE
USB Host MSC v1.1.1

### PR DESCRIPTION
**Pull request by tore-espressif**
# Change description

* Correctly handle empty string descriptors
* Move data buffer to DMA capable memory, if non-capable memory is passed to the driver

PTAL @Dazza0 @roma-jam 
cc @leeebo @lijunru-hub @Maldus512 @0727fqh @f1andrew

Closes https://github.com/espressif/esp-idf/issues/12163
Closes https://github.com/espressif/idf-extra-components/issues/202
Closes https://github.com/espressif/esp-idf/issues/12142

### CODE REVIEW

File: usb/usb_host_msc/include/usb/msc_host.h [hunk 1]
```diff
@@ -142,8 +142,15 @@ __attribute__((deprecated("use API from esp_private/msc_scsi_bot.h")));
...
+ * application needs to handle USB Host events itself.
+ * Do not use if MSC host install was made with create_background_task=true configuration
+ *
+ * @param[in]  timeout  Timeout in FreeRTOS tick
+ * @return
+ *     - ESP_OK:          All events handled
+ *     - ESP_ERR_TIMEOUT: No events handled within the timeout
+ *     - ESP_FAIL:        Event handling finished, driver uninstalled. You do not have to call this function further
  */
 esp_err_t msc_host_handle_events(uint32_t timeout_ms);
```
Comment (roma-jam) [inline usb/usb_host_msc/include/usb/msc_host.h:155]: "Timeout in FreeRTOS tick" but argument name `timeout_ms`
Looks like an error, isn't it?

File: usb/usb_host_msc/src/msc_host.c [hunk 2]
```diff
@@ -320,17 +322,34 @@ static bool is_mass_storage_device(uint8_t dev_addr)
...
 {
-    while (1) {
-        usb_host_client_handle_events(s_msc_driver->client_handle, portMAX_DELAY);
+    MSC_RETURN_ON_FALSE(s_msc_driver != NULL, ESP_ERR_INVALID_STATE);
 
-        if (s_msc_driver->end_client_event_handling) {
-            break;
-        }
+    ESP_LOGV(TAG, "USB MSC handling");
+    s_msc_driver->event_handling_started = true;
```
Comment (roma-jam) [inline usb/usb_host_msc/src/msc_host.c:332]: Here is the question: what if the MSC driver was configured with bg_task=false, but the `msc_host_handle_events` has never been called? 
Seems that `event_handling_started` flag will be staying false and semaphore does not let the `msc_host_uninstall` function be successfully completed. 

Does that make sense? :thinking:

File: usb/usb_host_msc/src/msc_host.c [hunk 3]
```diff
@@ -514,11 +537,20 @@ esp_err_t msc_host_write_sector(msc_host_device_handle_t device, size_t sector,
...
 {
-    MSC_RETURN_ON_FALSE(s_msc_driver != NULL, ESP_ERR_INVALID_STATE);
-
-    return usb_host_client_handle_events(s_msc_driver->client_handle, pdMS_TO_TICKS(timeout_ms));
+    if (dest == NULL) {
+        return;
+    }
+    if (src != NULL) {
+        size_t len = MIN((src->bLength - USB_STANDARD_DESC_SIZE) / 2, MSC_STR_DESC_SIZE - 1);
+        wcsncpy(dest, src->wData, len);
```
Comment (roma-jam) [inline usb/usb_host_msc/src/msc_host.c:547]: How do you deal with unaligned access warning here? 
Or you do not have any?

**Review by roma-jam — APPROVED** [review]
Left a few comments. 
Otherwise, LGTM. :clap:

File: usb/usb_host_msc/private_include/msc_common.h [hunk 4]
```diff
@@ -45,14 +45,15 @@ typedef struct msc_host_device {
  * @brief Trigger a BULK transfer to device: zero copy
  *
  * Data buffer ownership is transferred to the MSC driver and the application cannot access it before the transfer finishes.
- * The buffer must be DMA capable, as it is going to be accessed by USB DMA.
+ * This function is true zero-copy only if the passed data buffer is in DMA capable memory, as the USB Host uses DMA transfers.
+ * If the buffer is NOT DMA capable, an intermediate DMA capable buffer is allocated and used for the transfer, resulting in memory coping.
  *
- * This function significantly improves performance with usage of Virtual File System, which creates a intermediate buffer for each opened file.
- * The intermediate VFS buffer is then used for USB transfers too, which eliminates need of 2 large buffers and unnecessary copying of the data.
+ * This function significantly improves performance with usage of Virtual File System, which creates its own buffer for each opened file.
```
Comment (leeebo) [inline usb/usb_host_msc/private_include/msc_common.h:51]: AFAIK, `Virtual File System` does not maintain the buffer itself, the buffer is maintained by newlib with the `fopen` function, right?

Comment (tore-espressif) [reply usb/usb_host_msc/private_include/msc_common.h:51]: Yes, thank you! I'll update the docs

File: usb/usb_host_msc/src/msc_host.c (hunk 3 above)
Comment (tore-espressif) [reply usb/usb_host_msc/src/msc_host.c:547]: I suppressed this warning. It is not relevant as the descriptor is always in memory that supports unaligned access

File: usb/usb_host_msc/src/msc_host.c (hunk 2 above)
Comment (tore-espressif) [reply usb/usb_host_msc/src/msc_host.c:332]: Yes, it does. I will extend the tests to cover this situation

Comment (tore-espressif) [reply usb/usb_host_msc/src/msc_host.c:330]: Hi @roma-jam I extended the tests with the scenario you described, all tests passing!

Comment (roma-jam) [reply usb/usb_host_msc/src/msc_host.c:330]: Hi @tore-espressif ,

Yes, I did the same on HID and that also works.

**Pull request merged**